client/api/src/leaves.rs

```diff
@@ -99,27 +103,40 @@ where
 	}
 
 	/// Update the leaf list on import.
-	/// Returns a displaced leaf if there was one.
-	pub fn import(&mut self, hash: H, number: N, parent_hash: H) -> Option<ImportDisplaced<H, N>> {
-		// avoid underflow for genesis.
-		let displaced = if number != N::zero() {
-			let parent_number = Reverse(number.clone() - N::one());
-			let was_displaced = self.remove_leaf(&parent_number, &parent_hash);
-
-			if was_displaced {
-				Some(ImportDisplaced {
-					new_hash: hash.clone(),
-					displaced: LeafSetItem { hash: parent_hash, number: parent_number },
-				})
-			} else {
-				None
-			}
+	pub fn import(&mut self, hash: H, number: N, parent_hash: H) -> ImportOutcome<H, N> {
+		let number = Reverse(number);
+
+		let removed = if number.0 != N::zero() {
+			let parent_number = Reverse(number.0.clone() - N::one());
+			self.remove_leaf(&parent_number, &parent_hash).then(|| parent_hash)
+		} else {
+			None
+		};
+
+		self.insert_leaf(number.clone(), hash.clone());
+
+		ImportOutcome { inserted: LeafSetItem { hash, number }, removed }
+	}
+
+	/// Update the leaf list on removal.
+	/// Returns `None` if no modifications are applied.
+	pub fn remove(&mut self, hash: H, number: N, parent_hash: H) -> Option<RemoveOutcome<H, N>> {
+		let number = Reverse(number);
+
+		if !self.remove_leaf(&number, &hash) {
+			return None
+		}
+
+		let inserted = if self.storage.get(&number).is_none() && number.0 != N::zero() {
```

Comment:
Is this condition sufficient here?
Maybe I am missing something, but if we remove `12_2` in the following example, `11_2` should be added to the leave set right?
![CleanShot 2022-08-16 at 11 20 35](https://user-images.githubusercontent.com/5718005/184844878-1c5eed28-4396-4531-9246-46a6aea7feb4.png)

If I create this test, it fails:
```
#[test]
fn removal_breaks() {
		let mut set = LeafSet::new();
		set.import(10_1u32, 10u32, 0u32);
		set.import(11_1, 11, 10_1);
		set.import(12_1, 12, 11_1);

		set.import(11_2, 11, 10_1);
		set.import(12_2, 12, 11_2);

		let outcome = set.remove(12_2, 12, 11_2).unwrap();
		assert_eq!(outcome.removed.hash, 12_2);
		assert_eq!(outcome.inserted, Some(11_2));
}
```

Reply:
Good catch. 👍